### ISSUE TITLE
Map and path changes in `SystemJS.getConfig` after 0.20+

### ISSUE THREAD

**Issue by ArmorDarks** [systemjs/systemjs#1643]
Hi!

We're trying to tackle some related to `SystemJS.getConfig` issue here https://github.com/yoavniran/jest-jspm/issues/2.

Issue itself isn't relevant, don't waste time on it, but when I've explored why suddenly `jest-jspm` stopped to work after upgrading from SystemJS 0.19+ to SystemJS 0.20.11 Dev, I've noticed how paths in `SystemJS.getConfig()` changed:

![merge_2017-04-01_20-35-27](https://cloud.githubusercontent.com/assets/4460311/24581010/e16fcd98-171b-11e7-9d8e-a159fc2af87e.png)

(SystemJS v0.19.47 on the left, SystemJS 0.20.11 Dev on the right)

The most interesting part happens here:

![merge_2017-04-01_20-36-10](https://cloud.githubusercontent.com/assets/4460311/24581019/04d76a20-171c-11e7-9d3f-fa668c602b23.png)

It seemed like old `path` and `map` content have been more logical, since `path` contained possible mapping (npm and github) with full paths, and `map` simply used references to them.

But in SystemJS 0.20+ `map` contains full path and `path` seems to be redundant and not needed at all.

I just wanted to ensure that it should be like that and why exactly those changes has been made.

Thanks!

**Comment by guybedford** [comment]
Yes, this was done as a performance optimization to set the paths output of a mapped value to its final URL instead of doing it dynamically each normalization.

System.getConfig is a final mapping, not the "input" mapping information. This is the best we can do for System.getConfig though, because there's no reason for SystemJS to keep the input data around in memory otherwise, and should be releasing all these strings to the GC (which can take up a lot of space).

The perf optimization there is sticking, sorry it's caught an annoying case there - can you not get the configuration from the original source instead?

**Comment by ArmorDarks** [comment]
Thanks for reply!

I see. I understand the reason for this tradeoff.

I'd say that maybe it would worth to keep original values of `map` in another property, like `mapOrigin`, but I guess it will lead to a lot of memory wastage in large projects (since maps can be very big).

> can you not get the configuration from the original source instead?

No need to excuse, this issue isn't related to particularly our issue and completely workaraundable. I just wanted to ensure that it isn't a bug, since it's slightly unexpected output, especially considering that `path` property values now not usable at all. 

> can you not get the configuration from the original source instead?

Hm, I'm slightly confused. What meant under getting original source?

We're using SystemJS with JSPM, thus loading config from `jspm.config.js`. Because of quite specific structure of `jspm.config.js` with global variables `SystemJS` and lack of any exports, I don't see any other ways to load config except `getConfig()`. Or I'm missing something?

**Comment by guybedford** [comment]
It may be worth hooking the System.config function then if you can:

system-config-hook.js (loaded after SystemJS, but before jspm.config.js)
```javascript
var originalConfig;
var systemCfg = System.config;
System.config = function (cfg) {
  originalCfg = cfg;
  return systemCfg.apply(this, arguments);
};
```

It does get tricky though when config can be called multiple times then you need to follow the right extension rules, but something along those lines may well provide a path forward.

**Comment by ArmorDarks** [comment]
Thanks for suggestion!

Though, at first I was quite enthusiastic about offered solution, but then I realised that it seems more like a monkey patching `config` method and quite hacky, thus fragile. Especially considering that we already have `getConfig()`, but it simply doesn't do what it says and instead returns "compiled" config.

Unfortunately, there seems to be more problems with this monkey patching.

For instance, if we have following pretty much standard `jspm.config.js`

```js
SystemJS.config({
  baseURL: "assets/scripts",
  paths: {
    "npm:": "jspm_packages/npm/",
    "kotsu/": "source/scripts/"
  },
  devConfig: {
    "map": {
      "plugin-babel": "npm:systemjs-plugin-babel@0.0.13"
    }
  },
  transpiler: "plugin-babel",
  packages: {
    "kotsu": {
      "main": "main.js",
      "format": "esm",
      "meta": {
        "*.js": {
          "loader": "plugin-babel"
        }
      }
    }
  }
});

SystemJS.config({
  packageConfigPaths: [
    "npm:@*/*.json",
    "npm:*.json"
  ],
  map: {
    "jquery": "npm:jquery@3.1.0",
    "jump.js": "npm:jump.js@1.0.2"
  },
  packages: {}
});
```

Following approach will stumble us into curios issue.

I've slightly modified code:

```js
  SystemJS.config = function (cfg) {
    console.log('-------')
    console.log(cfg)
    console.log('-------')
    originalConfig = cfg
    return systemCfg.apply(this, arguments)
  }
```

Here what will output console:

```js
> node tests/utils/jest.js

-------
{ baseURL: 'assets/scripts',
  paths: { 'npm:': 'jspm_packages/npm/', 'kotsu/': 'source/scripts/' },
  devConfig:
   { map:
      { 'plugin-babel': 'npm:systemjs-plugin-babel@0.0.21',
        'systemjs-plugin-json': 'npm:systemjs-plugin-json@0.3.0' } },
  meta: { '*.json': { loader: 'systemjs-plugin-json' } },
  transpiler: 'plugin-babel',
  packages: { kotsu: { main: 'main.js', format: 'esm', meta: [Object] } } }
-------
-------
{ map:
   { 'plugin-babel': 'npm:systemjs-plugin-babel@0.0.21',
     'systemjs-plugin-json': 'npm:systemjs-plugin-json@0.3.0' } }
-------
-------
{ packageConfigPaths: [ 'npm:@*/*.json', 'npm:*.json' ],
  map: { jquery: 'npm:jquery@3.2.0' },
  packages: {} }
-------
```

Obviously this happens because `SystemJS.config()` called few times. Now you can imagine that there is no way here to identify which exactly `config()` call contains needed for us config except making some assumptions, which might be quite easily broken by any non-standard `jspm.config.js`.

**Comment by guybedford** [comment]
@ArmorDarks this is what I was referring to at the end of my comment. I think it may be easier here to rather than trying to detect the input configuration, simply normalize the output configuration to what jest-jspm is expecting.

You say the issue is the paths - in that case a normalization function could alter them to be what is expected, something along the lines of:

```javascript
normalizeConfigPaths(systemConfig) {
  Object.keys(systemConfig.paths).forEach(p => {
    if (systemConfig.paths[p].startsWith(systemConfig.baseURL))
      systemConfig.paths[p] = systemConfig.paths[p].substr(systemConfig.baseURL.length);
  });
}
```

**Comment by ArmorDarks** [comment]
> this is what I was referring to at the end of my comment.

Ah, sorry. I've seen your statement, but didn't realize at first that it was referencing exactly this issue. My bad.

> I think it may be easier here to rather than trying to detect the input configuration, simply normalize the output configuration to what jest-jspm is expecting.

That was my initial thought too.

I just wonder, how SystemJS detects needed input configuration? Maybe this functionality can be somehow exposed to 3rd-party libraries, like mentioned `jest-jspm`?

> You say the issue is the paths - in that case a normalization function could alter them to be what is expected, something along the lines of:

I think if there were a JavaScript Inquisition, they would already prepare beautiful fires for us :)

Well, I agree that this makes jobs done, but does it makes only me feel... like it's fundamentally wrong? Like it's sort of a fragile hack?

I mean, my first thought was to revert back original state too by simple replace, but it is strange that this need arises at first place.

Besides, all this means that we will need to mimick (in backwards) _whole routing logic of SystemJS_. Because paths may be different. BaseUrl might be in root, or it might be inside `browserConfig`, or inside `nodeConfig`, and many other potential caveats. This is evident path to breaking things easily.

As I already said, for now we can definitely go with this hacky way. But I think it just gives good illustration about current potential bottlenecks and things that could be improved.

I invite here maintainer of `jest-jspm`, @yoavniran, maybe he has something to say too.

--------------

I imagine if whole JSPM config would be simply exported, that issue won't raise at first place. Just import whatever you need and forget it:

```js
export const baseConfig = {
  baseURL: "assets/scripts",
  paths: {
    "npm:": "jspm_packages/npm/",
    "kotsu/": "source/scripts/"
  },
  devConfig: {
    "map": {
      "plugin-babel": "npm:systemjs-plugin-babel@0.0.13"
    }
  },
  transpiler: "plugin-babel",
  packages: {
    "kotsu": {
      "main": "main.js",
      "format": "esm",
      "meta": {
        "*.js": {
          "loader": "plugin-babel"
        }
      }
    }
  }
}

export const jspmConfig = {
  packageConfigPaths: [
    "npm:@*/*.json",
    "npm:*.json"
  ],
  map: {
    "jquery": "npm:jquery@3.1.0",
    "jump.js": "npm:jump.js@1.0.2"
  },
  packages: {}
}
```

```js
import baseConfig from './jspm.config.js'
```

This would solve a lot of issues, and, for instance, will help to avoid this "global-based variable" exposing of config.

But I realize that there is whole "story" with SystemJS loading config in browser, which isn't that easy. However, I wonder, maybe it could be simply initialized by SystemJS at first?

Like

```html
<script>
  SystemJS.init('./jspm.config.js).then(...)
</script>

After all, most part of SystemJS are already working by that time, they simply doesn't have a config file. But imports already should work, so SystemJS should be able to import config based on provided path.

I think it will make a lot of things cleaners and more straightforward, and also will reduce need to declare something globally.

What do you think?

**Comment by guybedford** [comment]
@ArmorDarks the above is a path forward, but yes not ideal.

Ideally I think there could be a nice util `jspm-config-extractor` that does something automatically and gives back the configuration. This could even be integrated into jspm core certainly.

**Comment by ArmorDarks** [comment]
> the above is a path forward, but yes not ideal.

You mean with substitution? Yeap, we seems to have no other options.

And what do you think about `export` in `jspm.config.js`? Theoretically, it won't be possible at all?

> Ideally I think there could be a nice util jspm-config-extractor that does something automatically and gives back the configuration. This could even be integrated into jspm core certainly.

Well, that would be great to have.

The main problem with such utility is that it should be closely tied to internals of SystemJS, based on its methods, otherwise I don't see how such utility will survive significant SystemJS updates.

In other words, ideally it should be one of SystemJS methods, especially considering that SystemJS already have everything needed to extract configs, so there is no reason to replicate most of its logic. It just need to have additional method which will spit it out before `getConfig()` will format everything inside of it. Something like `getRawConfig()`.

This way we can be sure that if something changes in SystemJS or JSPM, it will properly propagate to this utility method.

But dang, then it returns us to GC issue once again...

Hm, right know every possible "not hacky" solution I'm trying to think out stumbles into this issue with GC...

**Comment by guybedford** [comment]
@ArmorDarks configuration as a module is certainly a nicer pattern, if the implementation details can allow it. Certainly something to hope for in future.

**Comment by ArmorDarks** [comment]
Yeap, unfortunately I'm not aware of all implementation details. Though, yeah, that would be much better and right now something like `SystemJS.init('./jspm.config.js').then(...)` seems like very possible thing to me.

Anyway, as I understood, to solve our issue with `jest-jspm` we have to hack around. No other solutions, right?

Should I close this issue, or leave it as a reminder of potential domain, where things can be improved further?

**Comment by guybedford** [comment]
It's best to keep issues to immediate bugs and features, so yes let's close rather.

**Comment by ArmorDarks** [comment]
Yes, I understand. I'm quite upset with current ways of doing this, but still it's best we have for now. Thanks for your time!
